File: arc/cpp/runtime/constant/constant_test.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include <thread>

#include "gtest/gtest.h"

#include "x/cpp/test/test.h"

#include "arc/cpp/ir/ir.h"
#include "arc/cpp/runtime/constant/constant.h"
#include "arc/cpp/runtime/errors/errors.h"
#include "arc/cpp/runtime/state/state.h"

using namespace arc::runtime;

namespace {
node::Context make_context() {
    return node::Context{
        .elapsed = x::telem::SECOND,
        .mark_changed = [](const std::string &) {},
        .report_error = [](const x::errors::Error &) {},
        .activate_stage = [] {},
    };
}

struct TestSetup {
    arc::ir::IR ir;
    state::State state;

    TestSetup(const arc::types::Kind kind, const x::json::json &value):
        ir(build_ir(kind, value)),
        state(
            state::Config{.ir = ir, .channels = {}},
            arc::runtime::errors::noop_handler
        ) {}

    state::Node make_node() { return ASSERT_NIL_P(state.node("const")); }

private:
    static arc::ir::IR
    build_ir(const arc::types::Kind kind, const x::json::json &value) {
        arc::types::Param output_param;
        output_param.name = "output";
        output_param.type = arc::types::Type{.kind = kind};

        arc::types::Param value_param;
        value_param.name = "value";
        value_param.type = arc::types::Type{.kind = kind};
        value_param.value = value;

        arc::ir::Node ir_node;
        ir_node.key = "const";
        ir_node.type = "constant";
        ir_node.outputs.push_back(output_param);
        ir_node.config.push_back(value_param);

        arc::ir::Function fn;
        fn.key = "test";

        arc::ir::IR ir;
        ir.nodes.push_back(ir_node);
        ir.functions.push_back(fn);
        return ir;
    }
};
}

/// @brief Test that factory returns NOT_FOUND for non-constant node types.
TEST(ConstantFactoryTest, ReturnsNotFoundForWrongType) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    auto ir_node = setup.ir.nodes[0];
    ir_node.type = "not_constant";

    constant::Factory factory;
    ASSERT_OCCURRED_AS_P(
        factory.create(node::Config(setup.ir, ir_node, setup.make_node())),
        x::errors::NOT_FOUND
    );
}

/// @brief Test that factory creates a Constant node with valid configuration.
TEST(ConstantFactoryTest, CreatesConstantNode) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Factory factory;
    auto node = ASSERT_NIL_P(
        factory.create(node::Config(setup.ir, setup.ir.nodes[0], setup.make_node()))
    );
    ASSERT_NE(node, nullptr);
}

/// @brief Test that next() outputs the constant value on first call.
TEST(ConstantTest, NextOutputsValueOnFirstCall) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    ASSERT_NIL(node.next(ctx));

    auto checker = setup.make_node();
    const auto &output = checker.output(0);
    EXPECT_EQ(output->size(), 1);
    EXPECT_FLOAT_EQ(output->at<float>(0), 42.5f);
}

/// @brief Test that next() is a no-op on subsequent calls.
TEST(ConstantTest, NextNoOpsOnSubsequentCalls) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output = checker.output(0);
    output->set(0, 999.0f);

    ASSERT_NIL(node.next(ctx));
    EXPECT_FLOAT_EQ(output->at<float>(0), 999.0f);
}

/// @brief Test that reset() allows the value to be output again.
TEST(ConstantTest, ResetAllowsValueToBeOutputAgain) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output = checker.output(0);
    output->set(0, 999.0f);

    node.reset();
    node.next(ctx);

    EXPECT_FLOAT_EQ(output->at<float>(0), 42.5f);
}

/// @brief Test that float32 values are correctly cast and output.
TEST(ConstantTest, ValueIsCastToCorrectDataType_Float32) {
    TestSetup setup(arc::types::Kind::F32, 3.14f);
    constant::Constant node(setup.make_node(), 3.14f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output = checker.output(0);
    EXPECT_EQ(output->size(), 1);
    EXPECT_FLOAT_EQ(output->at<float>(0), 3.14f);
}

/// @brief Test that int64 values are correctly cast and output.
TEST(ConstantTest, ValueIsCastToCorrectDataType_Int64) {
    TestSetup setup(arc::types::Kind::I64, static_cast<int64_t>(12345));
    constant::Constant node(
        setup.make_node(),
        static_cast<int64_t>(12345),
        x::telem::INT64_T
    );

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output = checker.output(0);
    EXPECT_EQ(output->size(), 1);
    EXPECT_EQ(output->at<int64_t>(0), 12345);
}

/// @brief Test that uint8 values are correctly cast and output.
TEST(ConstantTest, ValueIsCastToCorrectDataType_U8) {
    TestSetup setup(arc::types::Kind::U8, static_cast<uint8_t>(255));
    constant::Constant node(
        setup.make_node(),
        static_cast<uint8_t>(255),
        x::telem::UINT8_T
    );

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output = checker.output(0);
    EXPECT_EQ(output->size(), 1);
    EXPECT_EQ(output->at<uint8_t>(0), 255);
}

/// @brief Test that is_output_truthy delegates to state.
TEST(ConstantTest, IsOutputTruthyDelegatesToState) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    EXPECT_TRUE(node.is_output_truthy("output"));
}

/// @brief Test that mark_changed is called on first next().
TEST(ConstantTest, MarkChangedCalledOnFirstNext) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    bool changed_called = false;
    std::string changed_param;
    auto ctx = make_context();
    ctx.mark_changed = [&](const std::string &param) {
        changed_called = true;
        changed_param = param;
    };

    node.next(ctx);

    EXPECT_TRUE(changed_called);
    EXPECT_EQ(changed_param, "output");
}

/// @brief Test that mark_changed is not called on subsequent next() calls.
TEST(ConstantTest, MarkChangedNotCalledOnSubsequentNext) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    int call_count = 0;
    ctx.mark_changed = [&](const std::string &) { call_count++; };

    node.next(ctx);
    node.next(ctx);

    EXPECT_EQ(call_count, 0);
}

/// @brief Test that timestamp is populated on first next().
TEST(ConstantTest, TimestampOutputOnFirstNext) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output_time = checker.output_time(0);
    EXPECT_EQ(output_time->size(), 1);
    EXPECT_GT(output_time->at<int64_t>(0), 0);
}

/// @brief Test that reset produces a new timestamp on subsequent next().
TEST(ConstantTest, ResetProducesNewTimestamp) {
    TestSetup setup(arc::types::Kind::F32, 42.5f);
    constant::Constant node(setup.make_node(), 42.5f, x::telem::FLOAT32_T);

    auto ctx = make_context();
    node.next(ctx);

    auto checker = setup.make_node();
    const auto &output_time = checker.output_time(0);
    const auto ts1 = output_time->at<int64_t>(0);

    node.reset();
    std::this_thread::sleep_for(std::chrono::milliseconds(1));
    node.next(ctx);

    const auto ts2 = output_time->at<int64_t>(0);
    EXPECT_GT(ts2, ts1);
}

File: arc/go/runtime/constant/constant_test.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package constant_test

import (
	"context"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"github.com/synnaxlabs/arc/graph"
	"github.com/synnaxlabs/arc/ir"
	"github.com/synnaxlabs/arc/runtime/constant"
	"github.com/synnaxlabs/arc/runtime/node"
	"github.com/synnaxlabs/arc/runtime/state"
	"github.com/synnaxlabs/arc/types"
	"github.com/synnaxlabs/x/query"
	"github.com/synnaxlabs/x/telem"
	. "github.com/synnaxlabs/x/testutil"
)

var ctx = context.Background()

var _ = Describe("Constant", func() {
	Describe("NewFactory", func() {
		It("Should create factory", func() {
			factory := constant.NewFactory()
			Expect(factory).ToNot(BeNil())
		})
	})

	Describe("Factory.Create", func() {
		var (
			factory node.Factory
			s       *state.State
		)
		BeforeEach(func() {
			factory = constant.NewFactory()
			g := graph.Graph{
				Nodes: []graph.Node{{Key: "const", Type: "constant"}},
				Functions: []graph.Function{{
					Key: "constant",
					Outputs: types.Params{
						{Name: ir.DefaultOutputParam, Type: types.I64()},
					},
				}},
			}
			analyzed, diagnostics := graph.Analyze(ctx, g, constant.SymbolResolver)
			Expect(diagnostics.Ok()).To(BeTrue())
			s = state.New(state.Config{IR: analyzed})
		})
		It("Should create node for constant type", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: 42}}},
				State: s.Node("const"),
			}
			Expect(MustSucceed(factory.Create(ctx, cfg))).ToNot(BeNil())
		})
		It("Should return NotFound for unknown type", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "unknown"},
				State: s.Node("const"),
			}
			Expect(factory.Create(ctx, cfg)).Error().To(MatchError(query.NotFound))
		})
		It("Should handle float64 value", func() {
			cfg := node.Config{
				Node: ir.Node{
					Type: "constant",
					Config: types.Params{{
						Name:  "value",
						Type:  types.F64(),
						Value: 3.14,
					}},
				},
				State: s.Node("const"),
			}
			Expect(MustSucceed(factory.Create(ctx, cfg))).ToNot(BeNil())
		})
		It("Should handle int value", func() {
			cfg := node.Config{
				Node: ir.Node{
					Type: "constant",
					Config: types.Params{{
						Name:  "value",
						Type:  types.I64(),
						Value: 100,
					}}},
				State: s.Node("const"),
			}
			Expect(MustSucceed(factory.Create(ctx, cfg))).ToNot(BeNil())
		})
		It("Should handle uint8 value", func() {
			cfg := node.Config{
				Node: ir.Node{
					Type: "constant",
					Config: types.Params{{
						Name:  "value",
						Type:  types.U8(),
						Value: uint8(255),
					}}},
				State: s.Node("const"),
			}
			Expect(MustSucceed(factory.Create(ctx, cfg))).ToNot(BeNil())
		})
	})

	Describe("Next", func() {
		var (
			s       *state.State
			factory node.Factory
			outputs []string
		)
		BeforeEach(func() {
			factory = constant.NewFactory()
			g := graph.Graph{
				Nodes: []graph.Node{{Key: "const", Type: "constant"}},
				Functions: []graph.Function{{
					Key: "constant",
					Outputs: types.Params{
						{Name: ir.DefaultOutputParam, Type: types.I64()},
					},
				}},
			}
			analyzed, diagnostics := graph.Analyze(ctx, g, constant.SymbolResolver)
			Expect(diagnostics.Ok()).To(BeTrue())
			s = state.New(state.Config{IR: analyzed})
			outputs = []string{}
		})

		It("Should emit output on Next with int value", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: 42}}},
				State: s.Node("const"),
			}
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(output string) {
				outputs = append(outputs, output)
			}})
			Expect(outputs).To(HaveLen(1))
			Expect(outputs[0]).To(Equal(ir.DefaultOutputParam))
		})

		It("Should set output data on Next", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: int64(100)}}},
				State: s.Node("const"),
			}
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			out := s.Node("const").Output(0)
			Expect(out.Len()).To(Equal(int64(1)))
		})

		It("Should set output time on Next", func() {
			cfg := node.Config{
				Node: ir.Node{
					Type:   "constant",
					Config: types.Params{{Name: "value", Type: types.F64(), Value: 3.14}},
				},
				State: s.Node("const"),
			}
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			outTime := s.Node("const").OutputTime(0)
			Expect(outTime.Len()).To(Equal(int64(1)))
			times := telem.UnmarshalSeries[telem.TimeStamp](*outTime)
			Expect(times[0]).To(BeNumerically(">", int64(0)))
		})

		It("Should handle float64 constant", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.F64(), Value: 2.718}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[float64](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			out := constNode.Output(0)
			vals := telem.UnmarshalSeries[float64](*out)
			Expect(vals[0]).To(Equal(2.718))
		})

		It("Should handle int32 constant", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I32(), Value: int32(42)}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[int32](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			out := constNode.Output(0)
			vals := telem.UnmarshalSeries[int32](*out)
			Expect(vals[0]).To(Equal(int32(42)))
		})

		It("Should handle uint8 constant", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.U8(), Value: uint8(255)}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[uint8](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			out := constNode.Output(0)
			vals := telem.UnmarshalSeries[uint8](*out)
			Expect(vals[0]).To(Equal(uint8(255)))
		})

		It("Should allow downstream nodes to read constant", func() {
			g := graph.Graph{
				Nodes: []graph.Node{
					{Key: "const", Type: "constant"},
					{Key: "sink", Type: "sink"},
				},
				Edges: []graph.Edge{
					{
						Source: ir.Handle{Node: "const", Param: ir.DefaultOutputParam},
						Target: ir.Handle{Node: "sink", Param: ir.DefaultInputParam},
					},
				},
				Functions: []graph.Function{
					{
						Key: "constant",
						Outputs: types.Params{
							{Name: ir.DefaultOutputParam, Type: types.I64()},
						},
					},
					{
						Key: "sink",
						Inputs: types.Params{
							{Name: ir.DefaultInputParam, Type: types.I64()},
						},
					},
				},
			}
			analyzed, diagnostics := graph.Analyze(ctx, g, constant.SymbolResolver)
			Expect(diagnostics.Ok()).To(BeTrue())
			s = state.New(state.Config{IR: analyzed})
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: int64(999)}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[int64](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			sink := s.Node("sink")
			recalc := sink.RefreshInputs()
			Expect(recalc).To(BeTrue())
			input := sink.Input(0)
			vals := telem.UnmarshalSeries[int64](input)
			Expect(vals[0]).To(Equal(int64(999)))
		})

		It("Should handle zero value constant", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: 0}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[int64](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			out := constNode.Output(0)
			vals := telem.UnmarshalSeries[int64](*out)
			Expect(vals[0]).To(Equal(int64(0)))
		})

		It("Should handle negative value constant", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: -42}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[int64](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(string) {}})
			out := constNode.Output(0)
			vals := telem.UnmarshalSeries[int64](*out)
			Expect(vals[0]).To(Equal(int64(-42)))
		})

		It("Should only emit once across multiple Next calls", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: int64(42)}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[int64](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(output string) {
				outputs = append(outputs, output)
			}})
			Expect(outputs).To(HaveLen(1))
			n.Next(node.Context{Context: ctx, MarkChanged: func(output string) {
				outputs = append(outputs, output)
			}})
			Expect(outputs).To(HaveLen(1))
		})

		It("Should emit again after Reset is called", func() {
			cfg := node.Config{
				Node:  ir.Node{Type: "constant", Config: types.Params{{Name: "value", Type: types.I64(), Value: int64(42)}}},
				State: s.Node("const"),
			}
			constNode := s.Node("const")
			*constNode.Output(0) = telem.NewSeriesV[int64](0)
			n, _ := factory.Create(ctx, cfg)
			n.Next(node.Context{Context: ctx, MarkChanged: func(output string) {
				outputs = append(outputs, output)
			}})
			Expect(outputs).To(HaveLen(1))
			n.Reset()
			n.Next(node.Context{Context: ctx, MarkChanged: func(output string) {
				outputs = append(outputs, output)
			}})
			Expect(outputs).To(HaveLen(2))
		})
	})

	Describe("SymbolResolver", func() {
		It("Should resolve constant symbol", func() {
			sym, ok := constant.SymbolResolver["constant"]
			Expect(ok).To(BeTrue())
			Expect(sym.Name).To(Equal("constant"))
		})
	})
})
